common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockPusher.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.spark.network.client.TransportClient;
+import org.apache.spark.network.shuffle.protocol.PushBlockStream;
+
+/**
+ * Similar to {@link OneForOneBlockFetcher}, but for pushing blocks to remote shuffle service to
+ * be merged instead of for fetching them from remote shuffle services. This is used by
+ * ShuffleWriter when the block push process is initiated. The supplied BlockFetchingListener
+ * is used to handle the success or failure in pushing each blocks.
+ */
+public class OneForOneBlockPusher {
+  private static final Logger logger = LoggerFactory.getLogger(OneForOneBlockPusher.class);
+  private static final ErrorHandler PUSH_ERROR_HANDLER = new ErrorHandler.BlockPushErrorHandler();
+
+  private final TransportClient client;
+  private final String appId;
+  private final String[] blockIds;
+  private final BlockFetchingListener listener;
+  private final Map<String, ManagedBuffer> buffers;
+
+  public OneForOneBlockPusher(
+      TransportClient client,
+      String appId,
+      String[] blockIds,
+      BlockFetchingListener listener,
+      Map<String, ManagedBuffer> buffers) {
+    this.client = client;
+    this.appId = appId;
+    this.blockIds = blockIds;
+    this.listener = listener;
+    this.buffers = buffers;
+  }
+
+  private class BlockPushCallback implements RpcResponseCallback {
+
+    private int index;
+    private String blockId;
+
+    BlockPushCallback(int index, String blockId) {
+      this.index = index;
+      this.blockId = blockId;
+    }
+
+    @Override
+    public void onSuccess(ByteBuffer response) {
+      // On receipt of a successful block push
+      listener.onBlockFetchSuccess(blockId, new NioManagedBuffer(ByteBuffer.allocate(0)));
+    }
+
+    @Override
+    public void onFailure(Throwable e) {
+      // Since block push is best effort, i.e., if we encountered a block push failure that's not
+      // retriable or exceeding the max retires, we should not fail all remaining block pushes.
+      // The best effort nature makes block push tolerable of a partial completion. Thus, we only
+      // fail the block that's actually failed. Not that, on the RetryingBlockFetcher side, once
+      // retry is initiated, it would still invalidate the previous active retry listener, and
+      // retry all outstanding blocks. We are preventing forwarding unnecessary block push failures
+      // to the parent listener of the retry listener. The only exceptions would be if the block
+      // push failure is due to block arriving on the server side after merge finalization, or the
+      // client fails to establish connection to the server side. In both cases, we would fail all
+      // remaining blocks.
+      if (PUSH_ERROR_HANDLER.shouldRetryError(e)) {
+        String[] targetBlockId = Arrays.copyOfRange(blockIds, index, index + 1);
+        failRemainingBlocks(targetBlockId, e);
+      } else {
+        String[] targetBlockId = Arrays.copyOfRange(blockIds, index, blockIds.length);
```

Comment:
The listener in `failRemainingBlocks` is the `RetryingBlockFetchListener`.
Here, if the exception is something that will happen for every block push, i.e., blocking being too late or we have a client side connection issue, all remaining blocks will fail.
By eagerly failing all remaining blocks, it could potentially speed up processing on the client side.

This is similar to how `OneForOneBlockFetcher` handles things.
The difference is that `OneForOneBlockFetcher` always eagerly fail all remaining blocks, while `OneForOneBlockPusher` does not.

Reply:
> Here, if the exception is something that will happen for every block push, i.e., blocking being too late or we have a client side connection issue, all remaining blocks will fail.

Would it have any side effect by calling `failRemainingBlocks` for multiple times for the same block in this case?

I mean, we may call `failRemainingBlocks` for multiple times like `failRemainingBlocks(Array(0, 1, ..., 4), e)`, `failRemainingBlocks(Array(1, ..., 4), e)`, ..., `failRemainingBlocks(Array(4), e)`.

Reply:
Each time we call `failRemainingBlocks`, if the block transfer is retriable, then `RetryingBlockFetchListener` will initiate the retry.
As part of that, it will retry all remaining block transfers and also change `currentListener`, so it would ignore any callback invocation for blocks from the previous attempt.
For example, once we invoke `failRemainingBlocks(Array(1, 2, 3, 4), e)`, block 1, 2, 3, 4 will be retired if e is retriable.
If later on, `failRemainingBlocks(Array(2, 3, 4), e)` is invoked, it means the transfer of block 2 during the retry encountered an issue, and thus we need to retry transferring block 2, 3, 4 a 2nd time.
Each invocation is for a unique attempt of transferring the blocks.

Reply:
I see. Thanks for the explanation.

But what's the case for the non-retriable listener or when `e` is not retriable? Seems like it may print duplicate logs for the same block. This's acceptable if the real `onBlockFetchFailure` operation is idempotent, e.g. `outstandingBlocksIds.remove(blockId)` for the `RetryingBlockFetchListener`'s case. But it's still good if we can improve it as a followup.

Reply:
I see. Thanks for the explanation.

But what's the case for the non-retriable listener or when `e` is not retriable? Seems like it may print duplicate logs for the same block. This's acceptable if the real `onBlockFetchFailure` operation is idempotent, e.g. `outstandingBlocksIds.remove(blockId)` for the `RetryingBlockFetchListener`'s case. But it's still good if we can improve it as a followup.

Reply:
If it's not retriable, the exception will then be surfaced to the parent listener, which will actually handle the failure.
The log will only be printed for the exceptions that are not retriable, and thus it should only print the log once per block by `RetryingBlockFetchListener`.
It will however print the log for the entire batch of blocks in the invocation of `failRemainingBlocks` with a non-retriable exception though.
This is the current behavior of Spark for block fetch failure as well.

In terms of parent listener, `RetryingBlockFetchListener` already has the mechanism in place to ensure it only surface the exception to the parent listener once per block, through its tracking of the current active listener.